File: frontend/src/api/schemas.ts
/**
 * Zod Runtime Validation Schemas for API Responses
 *
 * This module provides runtime validation for all API responses
 * to ensure type safety beyond compile-time TypeScript checks.
 *
 * All schemas use .strict() to catch unexpected properties.
 *
 * @module api/schemas
 * @since Phase 5.8 - Runtime Validation
 */

import { z } from 'zod';

// ============================================
// Helper for parsing Zod errors to ApiError format
// ============================================

/**
 * Formats a ZodError into a human-readable string
 */
export function formatZodError(error: z.ZodError): string {
  return error.issues
    .map((issue) => `${issue.path.join('.')}: ${issue.message}`)
    .join('; ');
}

// ============================================
// Base/Shared Schemas
// ============================================

/**
 * Schema for ConversationMessage (used in LoadResponse)
 */
const ConversationMessageSchema = z
  .object({
    type: z.enum(['player', 'narrator', 'matthew', 'tom']),
    text: z.string(),
    timestamp: z.number(),
  })
  .strict();


// ============================================
// Phase 1: Core Investigation Schemas
// ============================================

/**
 * Schema for InvestigateResponse
 * Runtime validation for POST /api/investigate
 */
export const InvestigateResponseSchema = z
  .object({
    narrator_response: z.string(),
    new_evidence: z.array(z.string()),
    evidence_names: z.record(z.string(), z.string()).optional(),
    already_discovered: z.boolean(),
    location_changed: z.string().optional(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


/**
 * Schema for SaveResponse
 * Runtime validation for POST /api/save
 */
export const SaveResponseSchema = z
  .object({
    success: z.boolean(),
    message: z.string().optional(),
  })
  .strict();


/**
 * Schema for LoadResponse
 * Runtime validation for GET /api/load/{case_id}
 */
export const LoadResponseSchema = z
  .object({
    case_id: z.string(),
    current_location: z.string(),
    discovered_evidence: z.array(z.string()),
    visited_locations: z.array(z.string()),
    conversation_history: z.array(ConversationMessageSchema).nullish(),
    narrator_verbosity: z.enum(['concise', 'storyteller', 'atmospheric']).optional(),
    language: z.string().optional(),
  })
  .strict();


/**
 * Schema for EvidenceResponse
 * Runtime validation for GET /api/evidence
 */
export const EvidenceResponseSchema = z
  .object({
    case_id: z.string(),
    discovered_evidence: z.array(z.string()),
  })
  .strict();


/**
 * Schema for EvidenceDetails
 * Runtime validation for GET /api/evidence/{id}
 */
export const EvidenceDetailsSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    location_found: z.string(),
    description: z.string(),
    type: z.string().optional(),
  })
  .strict();


/**
 * Schema for LocationResponse
 * Runtime validation for GET /api/case/{case_id}/location/{location_id}
 */
export const LocationResponseSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    description: z.string(),
    surface_elements: z.array(z.string()),
    witnesses_present: z.array(z.string()).nullish(),
  })
  .strict();


// ============================================
// Phase 2: Witness Schemas
// ============================================

/**
 * Schema for WitnessConversationItem
 */
const WitnessConversationItemSchema = z
  .object({
    question: z.string(),
    response: z.string(),
    timestamp: z.string(),
    trust_delta: z.number().optional(),
  })
  .strict();


/**
 * Schema for WitnessInfo
 * Runtime validation for GET /api/witness/{id} and GET /api/witnesses
 */
export const WitnessInfoSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    personality: z.string().nullable().optional(),
    trust: z.number(),
    conversation_history: z.array(WitnessConversationItemSchema).nullish(),
    secrets_revealed: z.array(z.string()),
    image_url: z.string().nullable().optional(),
  })
  .strict();


/**
 * Schema for InterrogateResponse
 * Runtime validation for POST /api/interrogate
 */
export const InterrogateResponseSchema = z
  .object({
    response: z.string(),
    trust: z.number(),
    trust_delta: z.number(),
    secrets_revealed: z.array(z.string()),
    secret_texts: z.record(z.string(), z.string()),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


/**
 * Schema for PresentEvidenceResponse
 * Runtime validation for POST /api/present-evidence
 */
export const PresentEvidenceResponseSchema = z
  .object({
    response: z.string(),
    trust: z.number(),
    trust_delta: z.number(),
    secrets_revealed: z.array(z.string()),
    secret_texts: z.record(z.string(), z.string()),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


// ============================================
// Phase 3: Verdict Schemas
// ============================================

/**
 * Schema for Fallacy
 */
const FallacySchema = z
  .object({
    name: z.string(),
    description: z.string(),
    example: z.string().optional(), // Backend: example: str = "" (has default)
  })
  .strict();


/**
 * Schema for MentorFeedbackData
 */
const MentorFeedbackDataSchema = z
  .object({
    analysis: z.string(),
    fallacies_detected: z.array(FallacySchema),
    score: z.number(),
    quality: z.string(),
    critique: z.string(),
    praise: z.string(),
    hint: z.string().nullable(), // Backend: hint: str | None = None (nullable)
  })
  .strict();


/**
 * Schema for DialogueLine
 */
const DialogueLineSchema = z
  .object({
    speaker: z.string(),
    text: z.string(),
    tone: z.string().optional(),
  })
  .strict();


/**
 * Schema for ConfrontationDialogueData
 */
const ConfrontationDialogueDataSchema = z
  .object({
    dialogue: z.array(DialogueLineSchema),
    aftermath: z.string(),
  })
  .strict();


/**
 * Schema for SubmitVerdictResponse
 * Runtime validation for POST /api/submit-verdict
 */
export const SubmitVerdictResponseSchema = z
  .object({
    correct: z.boolean(),
    attempts_remaining: z.number(),
    case_solved: z.boolean(),
    mentor_feedback: MentorFeedbackDataSchema,
    confrontation: ConfrontationDialogueDataSchema.nullable(),
    reveal: z.string().nullable(),
    wrong_suspect_response: z.string().nullable(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


// ============================================
// Phase 3.5: Briefing Schemas
// ============================================

/**
 * Schema for TeachingChoice
 */
const TeachingChoiceSchema = z
  .object({
    id: z.string(),
    text: z.string(),
    response: z.string(),
  })
  .strict();


/**
 * Schema for TeachingQuestion
 */
const TeachingQuestionSchema = z
  .object({
    prompt: z.string(),
    choices: z.array(TeachingChoiceSchema),
    concept_summary: z.string(),
  })
  .strict();


/**
 * Schema for CaseDossier
 */
const CaseDossierSchema = z
  .object({
    title: z.string(),
    victim: z.string(),
    location: z.string(),
    time: z.string(),
    status: z.string(),
    synopsis: z.string(),
  })
  .strict();


/**
 * Schema for BriefingContent
 * Runtime validation for GET /api/briefing/{case_id}
 */
export const BriefingContentSchema = z
  .object({
    case_id: z.string(),
    dossier: CaseDossierSchema,
    teaching_questions: z.array(TeachingQuestionSchema),
    transition: z.string().optional(),
    briefing_completed: z.boolean(),
    // Deprecated fields (optional for backward compatibility)
    case_assignment: z.string().optional(),
    teaching_question: TeachingQuestionSchema.optional(),
    rationality_concept: z.string().optional(),
    concept_description: z.string().optional(),
  })
  .strict();


/**
 * Schema for BriefingQuestionResponse
 * Runtime validation for POST /api/briefing/{case_id}/question
 */
export const BriefingQuestionResponseSchema = z
  .object({
    answer: z.string(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


/**
 * Schema for BriefingCompleteResponse
 * Runtime validation for POST /api/briefing/{case_id}/complete
 *
 * Backend returns {success, updated_state} after persisting briefing_completed flag.
 */
export const BriefingCompleteResponseSchema = z
  .object({
    success: z.boolean(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


// ============================================
// Phase 4: Matthew spirit companion schemas
// ============================================

/**
 * Schema for MatthewTriggerType
 */
const MatthewTriggerTypeSchema = z.enum([
  'helpful',
  'misleading',
  'self_aware',
  'dark_humor',
  'emotional',
]);

/**
 * Schema for MatthewTrigger
 * Runtime validation for POST /api/case/{case_id}/matthew/triggers/check
 */
export const MatthewTriggerSchema = z
  .object({
    id: z.string(),
    text: z.string(),
    type: MatthewTriggerTypeSchema,
    tier: z.union([z.literal(1), z.literal(2), z.literal(3)]),
  })
  .strict();


/**
 * Schema for MatthewResponse
 * Runtime validation for Matthew LLM endpoints (auto-comment and direct chat)
 */
export const MatthewResponseSchema = z
  .object({
    text: z.string(),
    mode: z.string(),
    trust_level: z.number(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


// ============================================
// Phase 5.2: Location Management Schemas
// ============================================

/**
 * Schema for LocationInfo
 * Runtime validation for GET /api/case/{case_id}/locations
 */
export const LocationInfoSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    type: z.string(),
  })
  .strict();


/**
 * Schema for ChangeLocationResponse
 * Runtime validation for POST /api/case/{case_id}/change-location
 */
export const ChangeLocationResponseSchema = z
  .object({
    success: z.boolean(),
    location: z
      .object({
        id: z.string(),
        name: z.string(),
        description: z.string(),
        surface_elements: z.array(z.string()).optional(),
        witnesses_present: z.array(z.string()).optional(),
      })
      .strict(),
    message: z.string().optional(),
    updated_state: z.record(z.string(), z.unknown()).optional(),
  })
  .strict();


// ============================================
// Phase 5.3: Save/Load System Schemas
// ============================================

/**
 * Schema for SaveSlotMetadata
 */
const SaveSlotMetadataSchema = z
  .object({
    slot: z.string(),
    case_id: z.string(),
    timestamp: z.string().nullable(),
    location: z.string(),
    evidence_count: z.number(),
    witnesses_interrogated: z.number().optional(),
    progress_percent: z.number().optional(),
    version: z.string(),
  })
  .strict();


/**
 * Schema for SaveSlotsListResponse
 * Runtime validation for GET /api/case/{case_id}/saves/list
 */
export const SaveSlotsListResponseSchema = z
  .object({
    case_id: z.string(),
    saves: z.array(SaveSlotMetadataSchema),
  })
  .strict();


/**
 * Schema for SaveSlotResponse
 * Runtime validation for POST /api/save (with slot)
 */
export const SaveSlotResponseSchema = z
  .object({
    success: z.boolean(),
    message: z.string(),
    slot: z.string().optional(),
  })
  .strict();


/**
 * Schema for DeleteSlotResponse
 * Runtime validation for DELETE /api/case/{case_id}/saves/{slot}
 */
export const DeleteSlotResponseSchema = z
  .object({
    success: z.boolean(),
    slot: z.string(),
    message: z.string().nullable(),
  })
  .strict();


// ============================================
// Phase 5.3.1: Landing Page Schemas
// ============================================

/**
 * Schema for ApiCaseMetadata
 */
const ApiCaseMetadataSchema = z
  .object({
    id: z.string(),
    title: z.string(),
    difficulty: z.enum(['beginner', 'intermediate', 'advanced']),
    description: z.string(),
  })
  .strict();


/**
 * Schema for CaseListResponse
 * Runtime validation for GET /api/cases
 */
export const CaseListResponseSchema = z
  .object({
    cases: z.array(ApiCaseMetadataSchema),
    count: z.number(),
    errors: z.array(z.string()).nullable().optional(),
  })
  .strict();


// ============================================
// Settings Schemas
// ============================================

/**
 * Schema for UpdateSettingsResponse
 * Runtime validation for POST /api/settings/update
 */
export const UpdateSettingsResponseSchema = z
  .object({
    success: z.boolean(),
    message: z.string(),
  })
  .strict();


// ============================================
// Phase 3: Reset Case Response Schema
// ============================================

/**
 * Schema for ResetResponse
 * Runtime validation for POST /api/case/{case_id}/reset
 */
export const ResetResponseSchema = z
  .object({
    success: z.boolean(),
    message: z.string(),
  })
  .strict();

File: backend/src/api/schemas.py
"""Pydantic request/response models for all API endpoints."""

from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

SaveSlotName = Literal["autosave", "slot_1", "slot_2", "slot_3"]

# Opaque ASCII request id for durable clients (Telegram gateway). Optional.
RequestId = Annotated[
    str | None,
    Field(
        max_length=128,
        pattern=r"^[\x21-\x7E]{1,128}$",
        description="Opaque ASCII idempotency key (max 128). Omit for legacy clients.",
    ),
]

# ============================================
# Investigation models
# ============================================


class InvestigateRequest(BaseModel):
    """Request for investigate endpoint."""

    player_input: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Player's action/input (max 1000 chars, ~250 tokens)",
    )
    case_id: str = Field(
        default="case_001",
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Case identifier",
    )
    location_id: str | None = Field(
        default=None,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Current location (optional, defaults to saved state or first location)",
    )
    slot: SaveSlotName = Field(
        default="autosave",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class StateDeltaResponse(BaseModel):
    """Lightweight player-state slice returned in SSE done payloads."""

    case_id: str
    current_location: str
    discovered_evidence: list[str] = Field(default_factory=list)
    visited_locations: list[str] = Field(default_factory=list)
    save_revision: int = 0


class InvestigateResponse(BaseModel):
    """Response from investigate endpoint."""

    narrator_response: str = Field(..., description="LLM narrator response")
    new_evidence: list[str] = Field(
        default_factory=list, description="Newly discovered evidence IDs"
    )
    evidence_names: dict[str, str] = Field(
        default_factory=dict, description="Evidence ID → display name map"
    )
    already_discovered: bool = Field(default=False, description="Was this already found?")
    location_changed: str | None = Field(
        default=None, description="New location ID if player moved via natural language"
    )
    updated_state: StateDeltaResponse | None = None


# ============================================
# Save/Load models
# ============================================


class SaveRequest(BaseModel):
    """Request for save endpoint."""

    state: dict[str, Any] = Field(..., description="Player state to save")
    slot: SaveSlotName = Field(
        default="autosave",
        description="Save slot to save state to",
    )


class SaveResponse(BaseModel):
    """Response from save endpoint."""

    success: bool
    message: str
    slot: str | None = None


class UpdateSettingsRequest(BaseModel):
    """Request to update player settings."""

    case_id: str = Field(
        default="case_001",
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Case identifier",
    )
    narrator_verbosity: str | None = Field(
        default=None,
        description="Narrator style: concise | storyteller | atmospheric",
    )
    language: str | None = Field(
        default=None,
        description="Game language: en | ru | fr | es | de | pt | zh | ja | ko",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class UpdateSettingsResponse(BaseModel):
    """Response from update settings endpoint."""

    success: bool
    message: str


class StateResponse(BaseModel):
    """Response containing player state."""

    case_id: str
    current_location: str
    discovered_evidence: list[str]
    visited_locations: list[str]
    conversation_history: list[dict[str, Any]] = []
    narrator_verbosity: str | None = None
    language: str | None = None


class ResetResponse(BaseModel):
    """Response for case reset endpoint."""

    success: bool
    message: str


class SaveSlotMetadata(BaseModel):
    """Metadata for a single save slot."""

    slot: str = Field(..., description="Slot identifier (slot_1, slot_2, slot_3, autosave)")
    case_id: str = Field(..., description="Case identifier (e.g., case_001)")
    timestamp: str | None = Field(None, description="Last save timestamp (ISO format)")
    location: str = Field(default="unknown", description="Current location in save")
    evidence_count: int = Field(default=0, description="Number of discovered evidence")
    witnesses_interrogated: int = Field(default=0, description="Number of witnesses interrogated")
    progress_percent: int = Field(default=0, ge=0, le=100, description="Progress percentage")
    version: str = Field(default="1.0.0", description="Save file version")


class SaveSlotsListResponse(BaseModel):
    """Response listing all save slots."""

    case_id: str
    saves: list[SaveSlotMetadata] = Field(default_factory=list)


class SaveSlotResponse(BaseModel):
    """Response from save/delete slot operations."""

    success: bool
    slot: str
    message: str | None = None


# ============================================
# Evidence models
# ============================================


class EvidenceDetailItem(BaseModel):
    """Evidence item with full metadata."""

    id: str
    name: str
    location_found: str
    description: str
    type: str


class EvidenceResponse(BaseModel):
    """Response from evidence endpoint."""

    case_id: str
    discovered_evidence: list[str]


class EvidenceDetailResponse(BaseModel):
    """Response with full evidence details."""

    case_id: str
    evidence: list[EvidenceDetailItem]


# ============================================
# Witness/Interrogation models
# ============================================


class InterrogateRequest(BaseModel):
    """Request for interrogate endpoint."""

    witness_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Witness identifier",
    )
    question: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Player's question (max 1000 chars, ~250 tokens)",
    )
    case_id: str = Field(
        default="case_001",
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Case identifier",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class InterrogateResponse(BaseModel):
    """Response from interrogate endpoint."""

    response: str = Field(..., description="Witness response")
    trust: int = Field(..., description="Current trust level (0-100)")
    trust_delta: int = Field(default=0, description="Change in trust from this question")
    secrets_revealed: list[str] = Field(
        default_factory=list, description="Secrets revealed in this response"
    )
    secret_texts: dict[str, str] = Field(
        default_factory=dict, description="Secret ID to full text description mapping"
    )
    updated_state: StateDeltaResponse | None = None


class PresentEvidenceRequest(BaseModel):
    """Request for present-evidence endpoint."""

    witness_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Witness identifier",
    )
    evidence_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Evidence to present",
    )
    case_id: str = Field(
        default="case_001",
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Case identifier",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class PresentEvidenceResponse(BaseModel):
    """Response from present-evidence endpoint."""

    response: str = Field(..., description="Witness response")
    trust: int = Field(..., description="Current trust level (0-100)")
    trust_delta: int = Field(default=0, description="Change in trust from this action")
    secrets_revealed: list[str] = Field(
        default_factory=list, description="Secrets revealed in this response"
    )
    secret_texts: dict[str, str] = Field(
        default_factory=dict, description="Secret ID to full text description mapping"
    )
    updated_state: StateDeltaResponse | None = None


class WitnessInfo(BaseModel):
    """Public witness information."""

    id: str
    name: str
    trust: int = Field(default=50, description="Current trust level")
    secrets_revealed: list[str] = Field(default_factory=list)
    conversation_history: list[dict[str, Any]] = Field(default_factory=list)
    personality: str | None = None


# ============================================
# Verdict models
# ============================================


class SubmitVerdictRequest(BaseModel):
    """Request for submit-verdict endpoint."""

    case_id: str = Field(
        default="case_001",
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Case identifier",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    accused_suspect_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Suspect ID being accused",
    )
    reasoning: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="Player's reasoning for accusation (max 2000 chars, ~500 tokens)",
    )
    evidence_cited: list[str] = Field(default_factory=list, description="Evidence IDs player cites")
    request_id: RequestId = None


class FallacyDetail(BaseModel):
    """Detailed fallacy information."""

    name: str
    description: str
    example: str = ""


class MentorFeedback(BaseModel):
    """Mentor feedback for verdict."""

    analysis: str
    fallacies_detected: list[FallacyDetail]
    score: int
    quality: str
    critique: str
    praise: str
    hint: str | None = None


class ConfrontationDialogue(BaseModel):
    """Confrontation dialogue after verdict."""

    dialogue: list[dict[str, str]]
    aftermath: str


class SubmitVerdictResponse(BaseModel):
    """Response from submit-verdict endpoint."""

    correct: bool
    attempts_remaining: int
    case_solved: bool
    mentor_feedback: MentorFeedback
    confrontation: ConfrontationDialogue | None = None
    reveal: str | None = None
    wrong_suspect_response: str | None = None
    updated_state: dict[str, Any] | None = None


# ============================================
# Briefing models
# ============================================


class TeachingChoice(BaseModel):
    """Single choice for teaching question."""

    id: str
    text: str
    response: str


class TeachingQuestion(BaseModel):
    """Teaching question with multiple choice answers."""

    prompt: str
    choices: list[TeachingChoice]
    concept_summary: str


class CaseDossier(BaseModel):
    """Structured case details for the briefing dossier."""

    title: str = "CLASSIFIED"
    victim: str
    location: str
    time: str
    status: str
    synopsis: str


class BriefingContent(BaseModel):
    """Briefing content from YAML."""

    case_id: str
    dossier: CaseDossier
    teaching_questions: list[TeachingQuestion]
    rationality_concept: str
    concept_description: str
    transition: str = ""
    briefing_completed: bool = False


class BriefingQuestionRequest(BaseModel):
    """Request for briefing question endpoint."""

    question: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Player's question for Graves (max 1000 chars, ~250 tokens)",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )


class BriefingQuestionResponse(BaseModel):
    """Response from briefing question endpoint."""

    answer: str = Field(..., description="Graves's response")
    updated_state: dict[str, Any] | None = None


class BriefingCompleteRequest(BaseModel):
    """Optional body for briefing complete. Query-only calls remain valid."""

    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class BriefingCompleteResponse(BaseModel):
    """Response from briefing complete endpoint."""

    success: bool
    updated_state: dict[str, Any] | None = None


# ============================================
# Spirit companion (Matthew) models
# ============================================


class InnerVoiceCheckRequest(BaseModel):
    """Request for inner voice check endpoint."""

    evidence_count: int = Field(..., ge=0, description="Current evidence count")


class InnerVoiceTriggerResponse(BaseModel):
    """Response from inner voice check endpoint (LEGACY YAML system)."""

    id: str = Field(..., description="Trigger ID")
    text: str = Field(..., description="Matthew's message text")
    type: str = Field(..., description="Trigger type (helpful/misleading/etc.)")
    tier: int = Field(..., ge=1, le=3, description="Trigger tier (1/2/3)")
    updated_state: dict[str, Any] | None = None


class MatthewAutoCommentRequest(BaseModel):
    """Request for Matthew auto-comment after evidence discovery."""

    is_critical: bool = Field(default=False, description="Force Matthew to comment?")
    last_evidence_id: str | None = Field(
        default=None,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Evidence just discovered",
    )


class MatthewChatRequest(BaseModel):
    """Request for direct Matthew conversation."""

    message: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Player's question to Matthew (max 1000 chars, ~250 tokens)",
    )


class MatthewResponseModel(BaseModel):
    """Matthew's response (LLM-powered)."""

    text: str = Field(..., description="Matthew's comment/response")
    mode: str = Field(
        ..., description="Response mode: 'auto', 'direct_chat', 'helpful', 'misleading'"
    )
    trust_level: int = Field(..., ge=0, le=100, description="Current trust level (0-100)")
    updated_state: dict[str, Any] | None = None


# ============================================
# LLM Configuration models
# ============================================


class VerifyKeyRequest(BaseModel):
    """Request to verify a user's API key."""

    provider: str = Field(..., description="Provider name (e.g. openrouter, anthropic)")
    api_key: str = Field(..., description="API key to verify")
    model: str | None = Field(None, description="Model to test with")


class VerifyKeyResponse(BaseModel):
    """Response from API key verification."""

    valid: bool
    error: str | None = None


class ModelInfo(BaseModel):
    """A recommended model option."""

    id: str
    name: str
    provider: str
    free: bool = False


# ============================================
# Case/Location models
# ============================================


class CaseListResponse(BaseModel):
    """Response from GET /api/cases endpoint."""

    cases: list[dict[str, Any]] = Field(default_factory=list, description="List of case metadata")
    count: int = Field(default=0, description="Number of valid cases")
    errors: list[str] | None = Field(default=None, description="Validation errors (if any)")


class LocationInfo(BaseModel):
    """Location metadata for LocationSelector."""

    id: str
    name: str
    type: str = "micro"


class ChangeLocationRequest(BaseModel):
    """Request for change-location endpoint."""

    location_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        pattern=r"^[a-zA-Z0-9_-]+$",
        description="Target location ID",
    )
    slot: str = Field(
        default="autosave",
        pattern=r"^[a-zA-Z0-9_]+$",
        description="Save slot to load/save state from",
    )
    request_id: RequestId = None


class ChangeLocationResponse(BaseModel):
    """Response from change-location endpoint."""

    success: bool
    location: dict[str, Any]
    updated_state: dict[str, Any] | None = None


# ============================================
# Telegram snapshot (internal gateway)
# ============================================


class TelegramWitnessSnapshot(BaseModel):
    """Canonical witness id + display name for Telegram clients."""

    id: str
    name: str
    description: str = ""


class TelegramEvidenceSnapshot(BaseModel):
    """Player-safe localized evidence text for Telegram casebook/buttons."""

    id: str
    name: str
    description: str = ""
    location_found: str = ""
    type: str = ""
    location_name: str = ""


class TelegramLocationSnapshot(BaseModel):
    """Player-safe localized location text for Telegram navigation."""

    id: str
    name: str
    description: str = ""


class TelegramSnapshotResponse(BaseModel):
    """Compact case snapshot for Telegram gateway. No secrets/solution."""

    case_id: str
    case_title: str = ""
    case_description: str = ""
    current_location: str
    current_location_view: TelegramLocationSnapshot | None = None
    available_locations: list[TelegramLocationSnapshot] = Field(default_factory=list)
    visited_locations: list[str] = Field(default_factory=list)
    discovered_evidence: list[str] = Field(default_factory=list)
    evidence_details: list[TelegramEvidenceSnapshot] = Field(default_factory=list)
    briefing_completed: bool = False
    language: str = "en"
    save_revision: int = 0
    available_witnesses: list[TelegramWitnessSnapshot] = Field(default_factory=list)
    verdict_attempts_remaining: int = 10
    case_solved: bool = False


# ============================================
# Telemetry models
# ============================================


class TelemetryEventRequest(BaseModel):
    """Request for telemetry event endpoint."""

    event_type: str = Field(..., max_length=64)
    case_id: str = Field(default="unknown", max_length=64, pattern=r"^[a-zA-Z0-9_-]+$")
    data: dict[str, Any] = Field(default_factory=dict)


class TelemetryErrorRequest(BaseModel):
    """Request for telemetry error endpoint."""

    error_type: str = Field(..., max_length=64)
    case_id: str = Field(default="unknown", max_length=64)
    message: str = Field(..., max_length=500)
    context: dict[str, Any] = Field(default_factory=dict)


class TelemetryResponse(BaseModel):
    """Response from telemetry endpoints."""

    ok: bool
